### DISCUSSION TITLE
TDD in Symfony (service removed when compiled)

### DISCUSSION THREAD

**Discussion by jdevinemt** [symfony/symfony#52895] [Q&A] [answered]
I've been using Symfony for a couple years now on a project and have just started writing tests with PHPUnit via the Symfony test-pack bundle. **I'm trying to use TDD for new source code, but have run into an annoying issue when writing integration tests for a service before I start using it.**

My integration tests extend the `KernelTestCase` class so that I can boot the kernel and retrieve the special test-version of the container via `static::getContainer()`, which allows retrieving private services. **However, because the service has not been used elsewhere in the code, aside from the tests, it gets removed during the compilation of the container.**

Removing unused services from the container makes total sense, but it would be really nice if it could know to check for usage in the configured tests directory when running tests (probably via the APP_ENV environment variable). **Is there a reason it does not do this? Is there some Symfony configuration I can set to achieve this?**

Currently I get around the issue by temporarily injecting the new service into a controller for a general test page I use. But it still does not resolve the issue that I cannot commit my changes until I actually use the service in my source code, since tests would break, meaning my code is in an undeployable state. Ideally, I would create a commit for creating the service, with a method or two, and then a second commit where I start using the service.

If anyone uses TDD and has suggestions on another way to look at this I would love to get your feedback.

**Comment by GromNaN** [comment, marked as answer]
You can make your service public, the time it's not injected into an other service. It will not be removed as unused.

**Reply by jdevinemt** [reply]
Haha, what a simple solution that I overlooked. Thank you. I'll just have to remember to change it back to private once it's used.
